x/ccv/provider/keeper/keeper.go

```diff
@@ -1236,6 +1236,39 @@ func (k Keeper) GetAllOptedIn(
 	return providerConsAddresses
 }
 
+func (k Keeper) HasToValidate(
+	ctx sdk.Context,
+	provAddr types.ProviderConsAddress,
+	chainID string,
+) (bool, error) {
+	// if the validator was sent as part of the packet in the last epoch, it has to validate
+	if k.IsConsumerValidator(ctx, chainID, provAddr) {
+		return true, nil
+	}
+
+	// if the validator was not part of the last epoch, check if the validator is going to be part of te next epoch
+	bondedValidators := k.stakingKeeper.GetLastValidators(ctx)
+	if topN, found := k.GetTopN(ctx, chainID); found && topN > 0 {
+		// in a Top-N chain, we automatically opt in all validators that belong to the top N
+		minPower := k.ComputeMinPowerToOptIn(ctx, chainID, bondedValidators, topN)
+		k.OptInTopNValidators(ctx, chainID, bondedValidators, minPower)
```

Comment:
hm, I think we should use a cached context here (is it expensive to create one?) just so this doesn't cause problems if it's called in another place in the future, particularly beacuse the method name sounds doesn't caution against calling it outside of queries